source/kraken/fill_disruption_from_chaos.cpp

```diff
@@ -402,25 +430,245 @@ struct functor_add_vj {
+  }
+  void register_vj_for_update(nt::FrequencyVehicleJourney* vj, nt::JourneyPattern* jp) {
+      vj_frequency_to_be_updated.push_back({vj, jp});
+  }
+
+  // update the adapated vj with the new journey pattern
+  // - update stop times
+  // - detach from its old jp
+  // - assign the new jp to the adapted vj
+  // - remove old jp from impact
+  // - add the adapted into new jp
+  template<typename VJ_T>
+  void complete_impl(const std::vector<std::pair<VJ_T*, nt::JourneyPattern*>>& vj_to_be_updated ) {
+
+      for (auto& vj_jp: vj_to_be_updated) {
+          auto& vj_ptr = vj_jp.first;
+          auto* new_jp = vj_jp.second;
+          auto* old_jp = vj_ptr->journey_pattern;
+
+          // We reconstruct a new stop time list whose stop points are not contained in the impacted
+          // stop points, then we move it to vj's stop time list
+          std::vector<nt::StopTime> stop_times;
+          for (auto& stop_time: vj_ptr->stop_time_list) {
+              const auto stop_point = stop_time.journey_pattern_point->stop_point;
+              auto it = boost::find_if(new_jp->journey_pattern_point_list,
+                      [stop_point](const nt::JourneyPatternPoint* jpp){ return jpp->stop_point == stop_point; });
+              // the stop_point of stop time does not belong to the jpp of the new_jp, we can remove it.
+              if (it == new_jp->journey_pattern_point_list.cend()) {
+                  continue;
+              }
+              stop_time.journey_pattern_point = *it;
+              stop_times.push_back(stop_time);
+          }
+          // calling the move constructor
+          vj_ptr->stop_time_list = std::move(stop_times);
+
+          auto& vj_list = get_vj_list_of_jp_helper<VJ_T>(old_jp);
+
+          auto it_old_vj = std::find_if(std::begin(vj_list), std::end(vj_list),
+                  [&vj_ptr](const std::unique_ptr<VJ_T>& old_vj ){
+                  return old_vj->uri == vj_ptr->uri;
+          });
+
+          assert(it_old_vj!= std::end(vj_list));
+
+          // Detach the vj from its jp and give it's ownership to a temporary unique_ptr
+          std::unique_ptr<VJ_T> old_vj_unique_ptr{it_old_vj->release()};
+          vj_list.erase(it_old_vj);
+
+          // Assign the new_jp to the vj
+          vj_ptr->journey_pattern = new_jp;
+
+          // We should also remove the old jp form impact's impacted_journey_patterns
+          boost::remove_erase_if(impact->impacted_journey_patterns, [&](const nt::JourneyPattern* jp){
+              return jp->uri == old_jp->uri;
+          });
+
+          // Now we can attach the vj to the  new jp
+          get_vj_list_of_jp_helper<VJ_T>(new_jp).push_back(std::move(old_vj_unique_ptr));
+
+      }
+  }
+
+  // update modified vj with its new jp to attach
+  void complete() {
+      complete_impl(vj_discrete_to_be_updated);
+      complete_impl(vj_frequency_to_be_updated);
+  };
+
+  bool is_impacted(const nt::VehicleJourney& vj) const{
```

Comment:
So here we only check that the vj is on the correct day, but not at the correct time?
If I understand correctly: application_periods are datetime, so it might matter, no?